### ISSUE TITLE
Make Local-then-Global Reductions possible in Chapel

### ISSUE THREAD

**Issue by LouisJenkinsCS** [chapel-lang/chapel#13218]
Currently, reductions are performed directly on the global operation, and so if we have N locales and M tasks, we have (N-1) * M remote reductions being performed; each of them performing an `execute_on` statement  to a single locale. 

https://github.com/chapel-lang/chapel/blob/470008e3df7392bab54381e1c7c6874104b7a49a/modules/internal/ChapelReduce.chpl#L64-L70

I would like for there to be a `param` flag or maybe a separate class that a local-first-then-global reduction can inherit from. As well it would be nice to be able to control _how_ the reduction is performed. For example, see `SumReduceScanOp`

https://github.com/chapel-lang/chapel/blob/470008e3df7392bab54381e1c7c6874104b7a49a/modules/internal/ChapelReduce.chpl#L140-L160

Which _must_ inherit from 

https://github.com/chapel-lang/chapel/blob/470008e3df7392bab54381e1c7c6874104b7a49a/modules/internal/ChapelReduce.chpl#L135-L138

Which is directly invoked (I.E not through an interface, or a method, but on the presumption that a field `l` exists that supports both `lock` and `unlock`), regardless of if the type could be reduced in a more efficient manner (also another possibility for a PR; let me know if I can within a 99.9% confidence margin, have such a contribution accepted, I will do my best to provide a patch upstream).

Example

```chpl
var x = + reduce 1..1024;
```

In the above, if we have M cores at our disposal, we'd still end up with M locks given the default granularity is 1 (unless you change this yourself), you'd end up acquiring the lock M times. However, a more efficient way would be to use `atomic` instead and `add` into that (with the benefit of RDMA atomics for global reductions).

Again, I volunteer myself to make these changes if I can get a 👍 that it will be accepted if I do, as I use reductions rather heavily in my code and any performance improvement would be beneficial.

**Comment by mppf** [comment]
@vasslitvinov - how is this related to ongoing reductions work?

**Comment by vasslitvinov** [comment]
This is surely a  big concern when reducing on a large system.

I believe that in 1.19 and currently on master the reduction tree matches the task-spawning tree. The task-spawning tree is defined by the parallel iterator that is invoked for the reduce expression.  When reducing a range, it is the range's standalone iterator. When reducing a block-distributed array, it is the block domain leader iterator. Etc. For example, as is the case for a Block array, if the parallel iterator has an outer coforall+on over locales, each locale gets its own globalOp object, in addition to the top-level globalOp. When each locale does a coforall over the cores, the partial result on each core is accumulated onto the globalOp on that locale.

@LouisJenkinsCS what kind of thing are you reducing where you care about how reductions are done?

Note that our planned next step with reductions is to switch to a different interface, see #9879. It will not by itself replace locking with atomic update, however.

**Comment by LouisJenkinsCS** [comment]
> I believe that in 1.19 and currently on master the reduction tree matches the task-spawning tree. The task-spawning tree is defined by the parallel iterator that is invoked for the reduce expression. 

I see, then it already does local-then-global reduction, I assumed that `globalOp` was literally the original variable you perform the reduction over. 

>  what kind of thing are you reducing where you care about how reductions are done?

Its more that I am going all-out on optimization to figure out where inefficiencies are and possibly point them out/fix them in the hopes of speeding up the rest of my code. I don't have anything in particular where reductions are too slow.

> Note that our planned next step with reductions is to switch to a different interface, see #9879. It will not by itself replace locking with atomic update, however.

Should I add to that issue or create an entirely new one asking for atomics to be used for`+ reduce` operations over `integral` types?

**Comment by vasslitvinov** [comment]
I see. Good to know.

> Should I add to that issue or create an entirely new one asking for atomics to be used for+ reduce operations over integral types?

A separate issue is appropriate. Once the new interface is in place, the decision of atomics vs. locks can be done separately from defining the accumulate and combine operations.

**Comment by LouisJenkinsCS** [comment]
Closing.
